### ISSUE TITLE
Long running process -> does not free memory. 

### ISSUE THREAD

**Issue by Ilkhom-S** [chronoxor/NetCoreServer#30]
Can you check out please. Service running after 2-3 days consuming about 2 gb ram . I think buffer is growing and not removing already readed data.

**Comment by Ilkhom-S** [comment]
There is 150 connections online  all time

**Comment by chronoxor** [comment]
For each session send buffer clears each time when the whole data was send to the client:

![image](https://user-images.githubusercontent.com/4397805/70434575-a346d900-1a95-11ea-9146-143f5290fe77.png)

Memory lead you described usually caused when you generate lots of data to send to clients and some slow client cannot handle the data on time and its send buffer grows.

You can handle or disconnect slow clients overriding 

```
protected override void OnSent(long sent, long pending)
{
    // Disconnect slow client with 10MB send buffer
    if (pending > 10485760)
        Disconnect();
}
```

**Comment by Ilkhom-S** [comment]
What about on receive buffer ?

**Comment by chronoxor** [comment]
If you handle receive data in your sessions without significant delays receive buffer will not grow a lot. The only case we reallocate receive buffer is when its capacity is full during the last receive phase:

![image](https://user-images.githubusercontent.com/4397805/70446333-faf13e80-1aad-11ea-9745-4ceec28bcf64.png)

In most cases receive buffer should not grow bigger than TCP window size.

**Comment by Ilkhom-S** [comment]
Can we do some option to clear receive buffer after receive handled by OnReceive method ?

**Comment by Ilkhom-S** [comment]
Cuz, I'm seeing that after 2-3 days working. Memory is about 2gb. I double check sessions. Sessions does not growing. It means GC collecting.

**Comment by chronoxor** [comment]
It will be good if you profile memory usage of your server so we'll know exactly the case and a place where memory leaks.

**Comment by chronoxor** [comment]
https://memprofiler.com/
http://www.jetbrains.com/dotmemory/

**Comment by Ilkhom-S** [comment]
ok, I'm not closing issue. I will try to find out, where memory leak is happening .

**Comment by Ilkhom-S** [comment]
Seems, TCPSession is not collected by GC  on disconnect it is removed from list but actually it's working at background. Can you make TCPSession disposable. So GC could lean faster.

**Comment by chronoxor** [comment]
Disposable will only help to free unmanged resources and other IDisposable objects. There is no way to say for GC to free byte[] memory buffers. Disconnected sessions with closed socket will be handled by GC automatically.

**Comment by chronoxor** [comment]
Turning on Server GC mode might also be helpful:
```
<PropertyGroup>
  <ServerGarbageCollection>true</ServerGarbageCollection>
</PropertyGroup>
```

**Comment by Ilkhom-S** [comment]
I made some changes to code. And restarted server. Let me check if it helps. I will feedback if any improvements

**Comment by Ilkhom-S** [comment]
Seems not helped, after 5 days work. Service using 1,5 gb ram

**Comment by batsword** [comment]
It's mostly a program problem, not NetCoreServer problem. 
I haven't had this problem in a few days

**Comment by willseward** [comment]
@Ilkhom-S I'm seeing almost the same issue. Do you see this with TLS and TCP or only TLS?
![2020-02-19 12_45_22 Microsoft Visual Studio](https://user-images.githubusercontent.com/5386568/74864849-cbfd1180-5315-11ea-9d5b-244fe633f30b.png)

**Comment by deviousasti** [comment]
If the receive buffers become sufficiently large, it will end up in the LOH.
Memory on the LOH can't be freed, only reallocated.
You can however try compacting the LOH.


```
GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
GC.Collect();
```

**Comment by Mohamed-Abubucker** [comment]
Hi @Ilkhom-S ,
me too facing the same issue. https://github.com/chronoxor/NetCoreServer/issues/46

for me, GC generation 2 is getting increased and causing memory leakage.

what about yours?

@chronoxor , please kindly help us to resolve this issue.

**Comment by deviousasti** [comment]
Can you post a detailed dump from a profiler about the breakup, especially allocation on the LOH.

**Comment by Mohamed-Abubucker** [comment]
Hi @deviousasti ,

this is the details of memory dump used by my netcoreserver tcp app.

```
Press p to pause, r to resume, q to quit.
    Status: Running

[System.Runtime]
    # of Assemblies Loaded                            80
    % Time in GC (since last GC)                      99
    Allocation Rate (Bytes / sec)              2,901,064
    CPU Usage (%)                                     12
    Exceptions / sec                                   0
    GC Heap Size (MB)                                648
    Gen 0 GC / sec                                     0
    Gen 0 Size (B)                            40,436,136
    Gen 1 GC / sec                                     0
    Gen 1 Size (B)                                    24
    Gen 2 GC / sec                                     0
    Gen 2 Size (B)                           732,575,352
    LOH Size (B)                                 582,488
    Monitor Lock Contention Count / sec                0
    Number of Active Timers                            2
    ThreadPool Completed Work Items / sec            136
    ThreadPool Queue Length                            0
    ThreadPool Threads Count                           9
    Working Set (MB)                                 827
```

**Comment by deviousasti** [comment]
LOH is not the problem. It seems okay at 500k.You need to take a heap snapshot and compare it with another to figure out what types are not GC’d.If you have access to the memory profiler in VS, use it. Can you report if any of the types in the diff are in the NetCoreServer assembly?  Sent from Mail for Windows 10 From: Mohamed AbubuckerSent: Saturday, February 29, 2020 7:43 PMTo: chronoxor/NetCoreServerCc: Asti; MentionSubject: Re: [chronoxor/NetCoreServer] Long running process -> does not free memory. (#30) Hi @deviousasti ,this is the details of memory dump used by my netcoreserver tcp app.Press p to pause, r to resume, q to quit.    Status: Running [System.Runtime]    # of Assemblies Loaded                            80    % Time in GC (since last GC)                      99    Allocation Rate (Bytes / sec)              2,901,064    CPU Usage (%)                                     12    Exceptions / sec                                   0    GC Heap Size (MB)                                648    Gen 0 GC / sec                                     0    Gen 0 Size (B)                            40,436,136    Gen 1 GC / sec                                     0    Gen 1 Size (B)                                    24    Gen 2 GC / sec                                     0    Gen 2 Size (B)                           732,575,352    LOH Size (B)                                 582,488    Monitor Lock Contention Count / sec                0    Number of Active Timers                            2    ThreadPool Completed Work Items / sec            136    ThreadPool Queue Length                            0    ThreadPool Threads Count                           9    Working Set (MB)                                 827—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by Mohamed-Abubucker** [comment]
@chronoxor , 

i need your help to resolve this memory leakage issue. i went live with your library, now it causing memory leakage. i am very much upset. 

kindly help me to resolve this.

**Comment by chronoxor** [comment]
Please provide minimal working example with your issue. We also have NetCoreServer solution in production for months online without memory-leaking issues.

**Comment by Mohamed-Abubucker** [comment]
Hi @chronoxor ,

I am just saving the received message into DB and writting the same into log file (i am using Nlog) and acknowledging the client with OK message.

my code should like this

```
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using NetCoreServer;

namespace core_tcp_5555
{
    public class SocketSession : TcpSession
    {

        private readonly ILogger<Runner> logger;

        public SocketSession(SocketServer server) : base(server)
        {
            logger = server.logger;
        }

        protected override void OnConnected()
        {
            Console.WriteLine($"Chat TCP session with Id {Id} connected!");
            Console.WriteLine($"Current Sessions Count {Server.ConnectedSessions}");
        }

        protected override void OnDisconnected()
        {

            Console.WriteLine($"Chat TCP session with Id {Id} and disconnected!");
            Console.WriteLine($"Current Sessions Count {Server.ConnectedSessions}");
        }

        protected override async void OnReceived(byte[] buffer, long offset, long size)
        {

            try
            {
                DateTime now = DateTime.Now;
                if (size > 0)
                {
                    string message = Encoding.ASCII.GetString(buffer, (int)offset, (int)size);
                    Console.WriteLine("Incoming: " + message);


                    // Here i am just saving the received message into database  
                    // await SaveMessageIntoDB(message);

                    byte[] data = Encoding.ASCII.GetBytes("OK");

                    _ = Task.Run(() =>
                    {
                        logger.LogInformation(message, "NEW MESSAGE RECEIVED");
                        SendAsync(data);
                    });
                }

            }
            catch (Exception e)
            {
                Console.WriteLine($"Error While OnReceived socket event - {e.Message + " " + e.StackTrace}");
                SendAsync($"ERROR - {e.Message + " " + e.StackTrace}");
            }
        }

        protected override void OnError(SocketError error)
        {

            Console.WriteLine($"Chat TCP session with Id {Id} caught an error with code {error}");
            Console.WriteLine($"Current Sessions Count {Server.ConnectedSessions}");
        }

        protected override void OnSent(long sent, long pending)
        {
            Disconnect();
        }
    }
}
```
i am running this same code in  2222, 5555 and 8888 port with pointing to different databases.
 
2222, and 8888 are working normally and it consumes 60mb to 200mb only. but the 5555 port running tcp server only causing the memory leakage issue.

i have tried many things, nothing worked for me. even i am disconnecting the client after acknowledged the client for message received.

still the memory leakage issue is only existing in 5555 port. 

kindly help me to resolve this ASAP.

**Comment by Mohamed-Abubucker** [comment]
@chronoxor ,

FYI, i am receiving 150K message per day and the RAM consumed size is getting increased gradually.

**Comment by Mohamed-Abubucker** [comment]
@chronoxor ,

i suspect that the client is keeping the connection always open (i am closing the connect only after i received one message from that particular client).

is there any possibilities to close the idle connections from server?
i checked the documentation, but no options like this.

kindly advise.

**Comment by chronoxor** [comment]
Keeping connection opened is OK, server is able to keep thousands of idle connections without any problems. Also you can extend server and create a watchdog timer inside and iterate through all `protected readonly Sessions` and disconnect idle ones.

Can you please try to get rid of the following fragment in your code:
```
                    _ = Task.Run(() =>
                    {
                        logger.LogInformation(message, "NEW MESSAGE RECEIVED");
                        SendAsync(data);
                    });
```

It looks that running new task for each incoming message is not important here. Just try to place logging and SendAsync() explicitly.

**Comment by Mohamed-Abubucker** [comment]
Hi @chronoxor , 

thank you for your advise, i have implemented your suggestion as well, but still the memory leakage issue not fixed.

**Comment by Ilkhom-S** [comment]
@Mohamed-Abubucker . I didn't find solution yet. I implement scheduler which restart service at night.  Seems library has issue with freeing resources. If connected clients are persistent, then memory not used much. But in my case, clients connects to server send messages and disconnects after 15 minutes. I have about 1K clients.

**Comment by Mohamed-Abubucker** [comment]
@Ilkhom-S ,

i am running the same code in multiple server in different ports.

only one port in each server has this memory leakage issue (5555 port).

i have renamed the project's solution, namespace, assembly name and everything to be unique for each application running in different ports.

i don't know why its behaving strange. apps running in some port are working fine without memory leakage and few are having memory leakage issues. same code-base with different assembly name and namespace.

my clients are mobile devices and i don't know they are keep opening the connection or closing the connection once the message sent.

anyway, from server, i am closing the connection after every message received and replied. my application is not multi cast or chat. its one to one communication. i mean each client will send the message to server and server will save the message into database then reply with OK and closes the connection.

could you help me in this how to schedule auto restart the app for everyday?

i mean sample code?

thanks in advance.

**Comment by Ilkhom-S** [comment]
If your service working under windows than you can just make new task for that. If you want to control from your code. Then write your service with topshelf library. And restart inside code.

**Comment by Mohamed-Abubucker** [comment]
Hi @Ilkhom-S ,

i am using Quartz cron scheduler in my application,

i used the below code to restart the app by every day 9am with with the help of Quartz library.

```
//Start process, friendly name is something like MyApp.exe (from current bin directory)
System.Diagnostics.Process.Start(System.AppDomain.CurrentDomain.FriendlyName);

//Close the current process
Environment.Exit(0);
```

i tested this and working fine. is this OK or anything changes needed?

**Comment by Ilkhom-S** [comment]
First you should close service and existing connections. Windows takes time to release memory used by service. After you can start new one

**Comment by Mohamed-Abubucker** [comment]
@Ilkhom-S 

Ok, i will do so.

**Comment by Ilkhom-S** [comment]
Can you implement to clear incoming bytes of OnReceive finished. It seems data is collected without any need.

**Comment by chronoxor** [comment]
There is no need to clear receive buffer after OnReceived() in method ProcessReceive() because after success received data and handle it receive loop will be launched again in TryReceive() method with _receiveBuffer starting from the offset 0:
```
                    _receiveEventArg.SetBuffer(_receiveBuffer.Data, 0, (int)_receiveBuffer.Capacity);
                    if (!Socket.ReceiveAsync(_receiveEventArg))
                        process = ProcessReceive(_receiveEventArg);
```

**Comment by chronoxor** [comment]
The only thing we have to do is to increase the receive buffer in case we are temporary too slow in processing receive data:
```
                // If the receive buffer is full increase its size
                if (_receiveBuffer.Capacity == size)
                    _receiveBuffer.Reserve(2 * size);
```

**Comment by Ilkhom-S** [comment]
I'm attaching profiler results. As you can see, buffer allocate near 600mb place in 30 min of work. 
![image](https://user-images.githubusercontent.com/446667/84023958-74f17a00-a9a2-11ea-94b6-2da5346445d7.png)

**Comment by Ilkhom-S** [comment]
Also, I noticed that u are not clearing receiveBuffer. 

` /// <summary>
        /// Clear send/receive buffers
        /// </summary>
        private void ClearBuffers()
        {
            lock (_sendLock)
            {
                // Clear send buffers
                _sendBufferMain.Clear();
                _sendBufferFlush.Clear();
                _sendBufferFlushOffset= 0;

                // Update statistic
                BytesPending = 0;
                BytesSending = 0;
            }
        }
`

is there any reason you are not clearing it ?

**Comment by Ilkhom-S** [comment]
Finally found issue. After session disconnected, inserted this code to clean unmanaged memory. When too many connections continuously connected and disconnected, unmanaged resources are not released. 

`  _receiveEventArg.Dispose();
   _sendEventArg.Dispose();
`
Hope you will fix this issue on .net framework 6 branch.

**Comment by Mohamed-Abubucker** [comment]
Hi All,

Thanks you and kindly fix this issue and release a new version as soon as possible
